picoquic/picoquic.h

```diff
@@ -708,23 +712,122 @@ void picoquic_delete_cnx(picoquic_cnx_t* cnx);
+ * The same structure is documented in the "bytes" and "length"
+ * parameter of the picoquic_callback_path_quality_changed callbacks,
+ * with the bytes parameter pointing to the structure, and the length
+ * indicating the structure length.
+ * 
+ * The quality changed callback only happens if the application
+ * subscribes to it using "picoquic_subscribe_to_quality_update" API
+ * for the path or for the connection, setting the "change" thresholds
+ * for the datarate and the rtt. The call applies to the specified
+ * path if the unique_path_id is valid, and to the whole connection
+ * if the unique_path_id is set to UINT64_MAX.
+ * The function call "picoquic_default_quality_update"
+ * can be used to set the default values of these parameters in
+ * the quic context.
+ * 
+ * If an error occurs, such as reference to an obsolete unique path id,
+ * all the functions return -1.
+ * 
+ * The call to "refresh the connection ID" will trigger a renewal of the connection
+ * ID used for sending packets on that path. This API is mostly used in test
+ * programs. By default, picoquic will attempt to renew a path connection ID 
+ * if that path resumes after a long silence: using
+ * a new connection ID in these conditions makes correlation of old and new
+ * connection data harder in case of NAT traversal.
  */
-int picoquic_probe_new_path(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
-    const struct sockaddr* addr_to, uint64_t current_time);
-int picoquic_abandon_path(picoquic_cnx_t* cnx, int path_id, uint64_t reason, char const* phrase);
 
-int picoquic_renew_connection_id(picoquic_cnx_t* cnx, int path_id);
+typedef struct st_picoquic_path_quality_t {
+    uint64_t pacing_rate;
+    uint64_t cwin;
+    uint64_t rtt;
+    uint64_t rtt_variant;
+    uint64_t rtt_min;
+    uint64_t sent;
+    uint64_t lost;
+    uint64_t bytes_in_transit;
+} picoquic_path_quality_t;
 
+int picoquic_probe_new_path(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
+    const struct sockaddr* addr_to, uint64_t current_time);
+int picoquic_probe_new_path_ex(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
+    const struct sockaddr* addr_to, int if_index, uint64_t current_time, int to_preferred_address);
+void picoquic_enable_path_callbacks(picoquic_cnx_t* cnx, int are_enabled);
+void picoquic_enable_path_callbacks_default(picoquic_quic_t* quic, int are_enabled);
+int picoquic_set_app_path_ctx(picoquic_cnx_t* cnx, uint64_t unique_path_id, void * app_path_ctx);
+int picoquic_abandon_path(picoquic_cnx_t* cnx, uint64_t unique_path_id, uint64_t reason, char const* phrase);
+int picoquic_get_path_quality(picoquic_cnx_t* cnx, uint64_t unique_path_id, picoquic_path_quality_t * quality);
+int picoquic_subscribe_to_quality_update(picoquic_cnx_t* cnx, uint64_t unique_path_id,
+    uint64_t pacing_rate_delta, uint64_t rtt_delta);
+void picoquic_default_quality_update(picoquic_quic_t* quic, uint64_t pacing_rate_delta, uint64_t rtt_delta);
```

Comment:
Feedback from @suhasHere: monitoring quality is important in both single path and multipath applications. It would be nice if there was an API to "get the quality on the current default path", so we could have a unified behavior. This would allow us to deprecate the API "picoquic_subscribe_pacing_rate_updates" and the callback "picoquic_callback_pacing_changed".

Reply:
This is addressed in the commit "create and test default path quality API", which:

* renames the previous `picoquic_subscribe_to_quality_update` to `picoquic_subscribe_to_quality_update_per_path`
* redefines `picoquic_subscribe_to_quality_update` to be for all the paths in the connection
* adds a new api `picoquic_get_default_path_quality`
* adds tests for the new API.